qiskit/transpiler/passes/calibration/rx_builder.py

```diff
@@ -0,0 +1,130 @@
+from qiskit.transpiler import Target
+from qiskit.circuit.library.standard_gates import RXGate
+from qiskit.exceptions import QiskitError
+
+
+class RXCalibrationBuilder(CalibrationBuilder):
+    """Add single-pulse RX calibrations that are bootstrapped from the SX calibration.
+
+    .. note::
+
+        Requirement: NormalizeRXAngles pass (one of the optimization passes).
+
+    It is recommended to place this pass in the post-optimization stage of a passmanager.
+    A simple demo:
+
+    .. code-block:: python
+
+       backend = FakeBelemV2()
+       pm = PassManager(RXCalibrationBuilder(backend.target))
+       qc = QuantumCircuit(1)
+       angles = [0.1, 0.2, 0.3, 0.4]
+       for angle in angles:
+         qc.rx(angle, 0)
+
+       # run the pass and check that new calibrations are generated
+       transpiled_circuit = pm.run(qc)
+       print(transpiled_circuit.calibrations["rx"])
+
+    References
+        * [1]: Gokhale et al. (2020), Optimized Quantum Compilation for
+          Near-Term Algorithms with OpenPulse.
+          `arXiv:2004.11205 <https://arxiv.org/abs/2004.11205>`
+    """
+
+    def __init__(
+        self,
+        target: Target = None,
+    ):
+        """Bootstrap single-pulse RX gate calibrations from the
+        (hardware-calibrated) SX gate calibration.
+
+        Args:
+            target (Target): Should contain a SX calibration that will be
+            used for bootstrapping RX calibrations.
+        """
+        from qiskit.transpiler.passes.optimization import NormalizeRXAngle
+
+        super().__init__()
+        self.target = target
+        self.already_generated = {}
+        self.requires = [NormalizeRXAngle(self.target)]
+
+    def supported(self, node_op: Instruction, qubits: list) -> bool:
+        """
+        Check if the calibration for SX gate exists.
+        """
+        return isinstance(node_op, RXGate) and self.target.has_calibration("sx", tuple(qubits))
+
+    def get_calibration(self, node_op: Instruction, qubits: list) -> Union[Schedule, ScheduleBlock]:
+        """
+        Generate RX calibration for the rotation angle specified in node_op.
+        """
+        # already within [0, pi] by NormalizeRXAngles pass
+        angle = node_op.params[0]
+
+        try:
+            angle = float(angle)
+        except TypeError as ex:
+            raise QiskitError("Target rotation angle is not assigned.") from ex
+
+        params = (
+            self.target.get_calibration("sx", tuple(qubits))
+            .instructions[0][1]
+            .pulse.parameters.copy()
+        )
+        new_rx_sched = _create_rx_sched(
+            rx_angle=angle,
+            channel=self.target.get_calibration("sx", tuple(qubits)).channels[0],
+            duration=params["duration"],
+            amp=params["amp"],
+            sigma=params["sigma"],
+            beta=params["beta"],
+        )
+
+        return new_rx_sched
+
+
+@lru_cache
+def _create_rx_sched(
```

Comment:
This is probably fine, but it makes me a little nervous. I would do more checking to confirm that the default calibration is a single Drag pulse.

Reply:
I added two tests in the `supported()` method to check if the default calibration is a single pulse and it's a DRAG pulse.
``` python
 def supported(self, node_op: Instruction, qubits: list) -> bool:
        """
        Check if the calibration for SX gate exists and it's a single DRAG pulse.
        """
        return (
            isinstance(node_op, RXGate)
            and self.target.has_calibration("sx", tuple(qubits))
            and (len(self.target.get_calibration("sx", tuple(qubits)).instructions) == 1)
            and isinstance(
                self.target.get_calibration("sx", tuple(qubits)).instructions[0][1].pulse, Drag
            )
        )
```

However, I can't decide whether we should allow non-DRAG pulses too or not.

- Pros for allowing non-DRAG pulses: Flexibility. This pass simply scales the amplitude of the default calibration, based on preliminary experiments that indicated no need to alter the DRAG coefficient. Since all other types of pulses (Gaussian, square, etc) have the amplitude parameter, it's technically possible to scale the amplitude of any type of pulses. 

- Cons for allowing non-DRAG pulses: The goal of this pass is to maximize the gate fidelity by decreasing gate time. The gate fidelity might not be optimal with non-DRAG pulses. Therefore, the user should probably try using DRAG pulses first, instead of using this pass.

Reply:
(I forgot to commit this separately. This accidentally became a part of the `Add a demo with a QuantumVolume circuit` commit.)

Reply:
I think it would be okay to check for `ScalableSymbolicPulse` and scale the `amp` parameter in that. Since this is not part of a preset passmanager and is a somewhat advanced concept, I think it is okay to assume that the user has some understanding of the pulse to be scaled. I was more concerned about a using doing something unusual like using a schedule with two pulses or with a variant of `Drag` and not realizing the schedule was being replaced with a single `Drag`.

Also, `isinstance(pulse, Drag)` is going to be deprecated. The recommendation is to use `isinstance(pulse, SymoblicPulse) and pulse.pulse_type == "Drag"` because `Drag` and the other symbolic pulses are being converted into factory functions that produce `SymbolicPulse`s instead of classes themselves. If you switch to checking for `ScalableSymbolicPulse` and scaling the `amp` this won't matter.

Reply:
Thank you @wshanks, I'm trying to cover the case where the original `sx` calibration is a `ScheduleBlock` with multiple types of instructions and multiple types of pulses, as you pointed out.
Could you help me building a new ScheduleBlock where everything else is the same but only the amplitude is halved?
I'm having trouble copying the alignment constraints and non-Play type of instructions.

This is an example of an unusually complicated sx schedule: 
``` python
d0 = pulse.DriveChannel(0)
with pulse.build() as complicated_sx_cal:
    pulse.play(Drag(duration=100, amp=0.1, sigma=40, beta=0.2, angle=10), d0)
    pulse.delay(50, d0)
    pulse.play(
        GaussianSquare(duration=130, amp=0.3, sigma=30, risefall_sigma_ratio=2, angle=30), d0
    )
```
![스크린샷 2023-09-20 오전 11 39 59](https://github.com/Qiskit/qiskit/assets/43903969/1ce4237e-8670-43c1-8e79-83066cd5fe50)


This is my attempt:
``` python
# goal: copy a ScheduleBlock, but halve the amplitude.

with pulse.build() as rx_cal:
    for i in range(len(complicated_sx_cal.instructions)):
        if isinstance(complicated_sx_cal.instructions[i][1], Play) and isinstance(
            complicated_sx_cal.instructions[i][1].pulse, ScalableSymbolicPulse
        ):
            # caveat: assumed that all the params are immutable (checked that Parameter is immutable)
            params = complicated_sx_cal.instructions[i][1].pulse.parameters.copy()
            halved_amp = params.pop("amp") * 0.5
            duration = params.pop("duration")
            angle = params.pop("angle", 0)
            pulse.play(
                ScalableSymbolicPulse(
                    complicated_sx_cal.instructions[i][1].pulse.pulse_type,
                    duration=duration,
                    amp=halved_amp,
                    angle=angle,
                    parameters=params,
                ),
                channel=complicated_sx_cal.channels[0]
            )  
        else:
            complicated_sx_cal.instructions[i][1]  #??
```

Questions:

- How do you copy a non-Play (ex. delay) instruction?
- How do you copy the alignment constraints in a ScheduleBlock? Or, would it be OK to assume that there won't be any alignment constraints for a gate calibration?

Reply:
When I print out the above `rx_cal`, the pulses seem to have the correct types (ex. `Drag` and `GaussianSquare`). However, I get `Pulse envelope expression is not assigned` errors when I try to draw the schedule.
`ScheduleBlock(Play(Drag(duration=100, sigma=40, beta=0.2, amp=0.05, angle=10), DriveChannel(0)), Play(GaussianSquare(duration=130, sigma=30, width=10.0, amp=0.15, angle=30), DriveChannel(0)), name="block25", transform=AlignLeft())`

Reply:
I suppose that "copying a ScheduleBlock but changing a certain pulse parameter" is routinely done for ZNE, so it must be implemented somewhere. But I can't seem to find it. Could you give me a hint, @nkanazawa1989 ? Thank you both in advance.

Reply:
Hmm, we can see what @nkanazawa1989 thinks. One option is not to try to handle such a complicated case, but to choose either to give an error or ignore it. A complicated schedule seems not too likely to come up, and either giving an error or ignoring the gate would avoid the case of assigning the wrong schedule (i.e. assigning a Drag with amplitude based on the first pulse amplitude if the schedule really has multiple instructions).

I think it's pretty tricky to walk through a `ScheduleBlock` and modify the pulses. I think I would not use the builder interface to do it. There is a `ScheduleBlock.replace` method that you could use like:

```python
new_sched = copy.deepcopy(sched)
for block in sched.blocks:
    if isinstance(block, Play) and isinstance(block.pulse, ScalableSymbolicPulse):
        new = Play(ScalableSymbolicPulse(pulse_type=block.pulse.pulse_type, ...<a lot of options to copy>), block.channel)
        new_sched.replace(block, new)
```

but that only covers the top level blocks. You would also need to check `isinstance(block, ScheduleBlock)` and recurse into the nested `ScheduleBlock`s and replace pulses in them and call `replace(block, new_scheduleblock)`.

Reply:
I see, I agree that it's reasonable to reject anything else than a single Drag pulse. The current implementation of supported() checks that.
https://github.com/Qiskit/qiskit/pull/10634/files#diff-5c2525ef082142486a945b9138611abe282cc048803c89059207112f8db192b8R103-R106
Thank you for the example on `replace()`.

Reply:
Copying a `ScalableSymbolicPulse` is a bit awkward. It should work if you copy all the input parameters:

```python
p = pulse.Drag(100, 0.5, 10, 0.1)
p2 = pulse.ScalableSymbolicPulse(
    pulse_type=p.pulse_type,
    duration=p.duration,
    amp=p.amp,
    angle=p.angle,
    parameters={k: v for k, v in p.parameters.items() if k not in ("amp", "angle", "duration")},
    limit_amplitude=p.limit_amplitude,
    envelope=p.envelope,
    constraints=p.constraints,
    valid_amp_conditions=p.valid_amp_conditions,
)
```

It would be nice to be able to copy the pulse and mutate the `amp`, but mutation is not part of the design. A `SymbolicPulse.copy_with_new_parameters()` method might be nice to have.

Reply:
Thank you for the discussion 😊 I updated the Drag check according to your suggestion. (In other words, the current version supports only the calibrations with a single Drag pulse)
As a next step, I will support calibrations with a single `ScalableSymbolicPulse`. 
The final step would be to support calibrations with multiple instructions (nested `ScheduleBlock`, non-Play instructions, etc). However, such calibrations seem unlikely to occur. Moreover, I may need to introduce some arbitrary rules to reject edge cases. Therefore, I would like to require in `supported()` that sx calibration should consist of a single `ScalableSymbolicPulse`.